processing/context_collector.go

```diff
@@ -0,0 +1,198 @@
+}
+
+// Context is a collection of JSON events that belong to a given flow.
+type Context []string
+
+// MakeContextCollector creates a new ContextCollector.
+func MakeContextCollector(shipper ContextShipper) *ContextCollector {
+	c := &ContextCollector{
+		Logger: log.WithFields(log.Fields{
+			"domain": "context",
+		}),
+		Cache:  cache.New(ContextCollectorDefaultTTL, ContextCollectorDefaultTTL),
+		Marked: make(map[string]struct{}),
+		i:      0,
+		Ship:   shipper,
+	}
+	return c
+}
+
+// Mark queues metadata for a given flow for forwarding, identified by its
+// flow ID.
+func (c *ContextCollector) Mark(flowID string) {
+	// when seeing an alert, just mark the flow ID as relevant
+	c.MarkLock.Lock()
+	c.Marked[flowID] = struct{}{}
+	c.MarkLock.Unlock()
+}
+
+// Consume processes an Entry, adding the data within to the internal
+// aggregated state
+func (c *ContextCollector) Consume(e *types.Entry) error {
+	var myC Context
+	// Some events, e.g. stats, have no flow ID set
+	if e.FlowID == "" {
+		return nil
+	}
+
+	cval, exist := c.Cache.Get(e.FlowID)
+	if exist {
+		// the 'flow' event always comes last, so we can use it as an
+		// indicator that the flow is complete and can be processed
+		if e.EventType == "flow" {
+			var isMarked bool
+			c.MarkLock.Lock()
+			if _, ok := c.Marked[e.FlowID]; ok {
+				isMarked = true
+			}
+			c.MarkLock.Unlock()
+			if isMarked {
+				c.StatsLock.Lock()
+				c.PerfStats.Flows++
+				c.PerfStats.Events += uint64(len(cval.(Context)))
+				for _, v := range cval.(Context) {
+					c.PerfStats.JSONBytes += uint64(len(v))
+				}
+				c.StatsLock.Unlock()
+				c.Ship(cval.(Context), c.Logger)
+				delete(c.Marked, e.FlowID)
+			}
+			c.Cache.Delete(e.FlowID)
+		} else {
+			myC = cval.(Context)
+			myC = append(myC, e.JSONLine)
+			c.Cache.Set(e.FlowID, myC, cache.DefaultExpiration)
+		}
+	} else {
+		if e.EventType != "flow" {
+			myC = append(myC, e.JSONLine)
+			c.Cache.Set(e.FlowID, myC, cache.DefaultExpiration)
+		}
+	}
+	c.i++
+	if c.i%100000 == 0 {
+		count := c.Cache.ItemCount()
+		c.Logger.WithFields(log.Fields{
+			"n": count,
+		}).Debugf("cache size after another 100k events")
+		c.i = 0
+	}
+	return nil
+}
+
+func (c *ContextCollector) runCounter() {
+	sTime := time.Now()
+	for {
+		time.Sleep(500 * time.Millisecond)
+		select {
+		case <-c.StopCounterChan:
+			close(c.StoppedCounterChan)
+			return
+		default:
+			if c.StatsEncoder == nil || time.Since(sTime) < c.StatsEncoder.SubmitPeriod {
+				continue
+			}
+			c.StatsEncoder.Submit(c.PerfStats)
+			c.StatsLock.Lock()
+			c.PerfStats.JSONBytes = 0
+			c.PerfStats.Flows = 0
+			c.PerfStats.Events = 0
+			sTime = time.Now()
+			c.StatsLock.Unlock()
+		}
+	}
+}
+
+// GetName returns the name of the handler
+func (c *ContextCollector) GetName() string {
+	return "Context collector"
+}
+
+// GetEventTypes returns a slice of event type strings that this handler
+// should be applied to
+func (c *ContextCollector) GetEventTypes() []string {
```

Comment:
Pointer receiver not needed.

Reply:
Does not hurt either, and since `ContextCollector` contains a mutex, making this a non-pointer receiver would cause:
```
go vet ./...
# github.com/DCSO/fever/processing
processing/context_collector.go:168:9: GetEventTypes passes lock by value: github.com/DCSO/fever/processing.ContextCollector contains sync.Mutex
```